Split setup process into atomic stages

sloansparger opened the pull request:
Closes #32768

### Description

Updates the setup process to call POST /api/setup after the language and user details step is completed then proceeds to make every step after completely atomic. This helps reduce the time the instance spends using the setup token as an auth strategy and tightens security. Consequently, we can no longer need to rely on some of the other setup endpoints as we already have endpoints for fully authenticated users (which the user becomes after they submit their details).

### Product callouts

- There's been some UX changes required to make this feature more modular / easier to ship. Since the main setup endpoint can only be called once, we can't allow users
to go back to the first two steps. 

- The other downside to this new approach is we have no "abandoned cart" functionality if the user doesn't complete the setup after the user details step. They will have provided as much information as needed to initialize their instance, so all further data will have to be collected from within the application.


### Summary of the list of updates make to FE/BE:

- make first two steps of setup flow an atomic step
- reduces setup options accepted by POST /api/setup
- removes POST /api/setup/validate endpoint to validate a db connection
- updates POST /api/database/validate endpoint to result with same response schema as POST /api/setup/validate did (preserves existing data as well)
- DB creation happens via POST /api/database
- FE relies on settings endpoints for things like allowing/disallowing tracking, saving license token, etc.
- removes POST /api/setup/token-check as we can rely on a 400 response to detect validity

### Demo

https://github.com/metabase/metabase/assets/7104357/5bbb1963-e7b4-4daa-80df-5928a7839e16


### How to verify
- Setup a new instance and test that setup flow is functionally identical to existing behaviors (except the aforementioned no going back)

### Checklist
- [x] Tests have been added/updated to cover changes in this PR

npretto committed:
adds /api/setup/token-check and add license-token to POST /api/setup (#38858)

* adds /api/setup/token-check and add license-token to POST /api/setup

* refactor tests

* license-token => license_token for consistency

* fixed test message

* fix nesting in test

npretto committed:
Add the step in the FE and pass the token to the setup api call (#38953)

* refactor: go back to saving the index for the step, to avoid having to use a selector that needs global state in the reducer

* Revert "refactor: go back to saving the index for the step, to avoid having to use a selector that needs global state in the reducer"

This reverts commit ee7e44f48788528e059c3f1f8816a73b675e082b.

* feat: adds UI and token api logic

* test structure for paid features

* initSetup action as workaround

* pass token to the /setup endpoint

* add jest tests for ee without token case

* add warning comment  on getIsPaidPlan

* copy old schema for better diff

* add license_token step seen event in the schema

* update e2e tests to make them pass

* should make e2e pass

* explicitly check that the license activation step is not rendered

* useStep hook

* refactor: simplify logic in step_number

* refactor: getInactiveStepTitle

* fix too many slashes in import

* refactor: rename selectThisStep to handleStepSelect

* use isSetupCompleted from useStep

* adds expect().toBeInTheDocument

* remove waitFor in favor of await findBy

npretto committed:
add submit snowplow event and increase the onboarding version in the events (#39151)

* trackLicenseTokenStepSubmitted

* adds a test and makes sure we're passing the token as undefined, not as null

* clean up

* remove waitFor

npretto committed:
goToNextStep in all thunks (#39103)

* Revert "initSetup action as workaround"

This reverts commit 0318a55b22b189af9e30d4a8116bacfb0365fcd2.

* goToNextStep thunk

sloansparger committed:
wip - submits user data earlier in setup process, updates other steps to call other endpoints making each step after user creation atomic

npretto committed:
Change the Next button to skip for the license activation step (#39292)

* change the Next button to be Skip

* trim token input

* fix e2e test

npretto committed:
Merge branch 'master' into license-activation-step

npretto committed:
fix quotes

johnswanson committed:
return details from `/api/database/validate`

sloansparger committed:
utilizes updates db validation endpoint, reverts some previous work that was not needed, creates db at end of setup instead of before, stops showing nav/side bars when the user completes the user creation step

sloansparger committed:
dispatch updateSettings action over calling endpoint

sloansparger committed:
integrate with updated validation endpoint response

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger committed:
removes going back to user creation in e2e tests

sloansparger committed:
fix unit tests

sloansparger committed:
fixes data usage unit test step

sloansparger committed:
defer token submission to avoid reloading setup experience

sloansparger committed:
clean up pass

sloansparger committed:
second cleanup pass

sloansparger committed:
last pass at clean up

sloansparger commented:
Remaining backend work
- remove unused fields in setup endpoint (you can see the used fields [here](https://github.com/metabase/metabase/pull/39287/files#diff-f0f1344935a4c128194840b010588c50a1260aa9dae823a4b8c51dee7e130284R104-R112))
- remove old setup db validation endpoint (`POST /api/setup/validate` - src/metabase/api/setup.clj:190)
- move `POST /api/setup/token-check` to something like `/api/token-check` and replace `setup/has-user-setup` clause with an admin check (src/metabase/api/setup.clj:355) - can we make sure that the response body returns the same validation data that `POST /api/setup` is doing now? 
- double check that you can't validate a database when unauthenticated (old endpoint was giving me 4XX status but still with a validation check body)

sloansparger committed:
fixes enterprise unit test

sloansparger committed:
fixes failures in setup/tests/common.unit.spec.tsx

sloansparger committed:
whack a mole

escherize committed:
remove setup/validate

escherize committed:
add superuser check to api/setup/token-check

escherize committed:
reduce input surface area for POST api/setup

sloansparger committed:
change snowplow assertion so e2e test can be debugged better...

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger committed:
test different snowplow schema for reporting tracking settings change

sloansparger committed:
removes duplicated tracking change

sloansparger committed:
fixes tracking button not toggling if network request to save fails

escherize committed:
remove test that tested db creation.

- we cannot create dbs with this endpoint anymore so we don't need to
  check.

sloansparger committed:
fixes test failing from make test setup state default to enabling tracking by default

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger commented on e2e/support/helpers/e2e-snowplow-helpers.js:
```diff
@@ -91,7 +91,7 @@ export const expectGoodSnowplowEvents = count => {
 
 export const expectNoBadSnowplowEvents = () => {
-  sendSnowplowRequest("micro/bad").its("body").should("have.length", 0);
+  sendSnowplowRequest("micro/bad").its("body").should("deep.equal", []);
```
make it easier to debug failures

sloansparger commented on frontend/src/metabase/selectors/app.ts:
```diff
@@ -20,6 +20,7 @@ export interface RouterProps {
 }
 
 const PATHS_WITHOUT_NAVBAR = [
+  /^\/setup/,
```
navbar and sidebar would appear when the setup endpoint was called, but the user was not finished with the setup process

escherize committed:
adds conenction_source qp to POST api/database/

- We were setting up users from an other endpoint, which would add
  `:source` as `:setup` in a call to `snowplow/track-event!`. But we are
  changing how setup works, and we will passin the connection_source=setup as
  a query param POST api/database/ so we don't lose that information

- defaults to :admin

escherize committed:
Merge branches 'ss/multi-req-setup' and 'ss/multi-req-setup' of github.com:metabase/metabase into ss/multi-req-setup

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger committed:
fix unrelated e2e test flakes

sloansparger committed:
last pass at PR clean up

escherize committed:
Don't test db creation on POST /api/setup

- We no longer create a database on POST /api/setup

escherize committed:
remove more tests that check for db creation at "startup"

sloansparger committed:
increase input delay for 30680-native-editor-buttons.cy.spec.js

sloansparger committed:
try to wait longer for 17514-ui-overlay.cy.spec.js failure

escherize committed:
fix tests in api/setup-test

escherize committed:
Merge remote-tracking branch 'origin/ss/multi-req-setup' into ss/multi-req-setup

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

escherize committed:
fixing more tests

escherize committed:
Merge branch 'ss/multi-req-setup' of github.com:metabase/metabase into ss/multi-req-setup

escherize committed:
remove test that's no longer needed

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger committed:
makes it to where user can not go back a step in setup, converts license input to formik, updates each step to submit its own data

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger commented on e2e/test/scenarios/native/reproductions/30680-native-editor-buttons.cy.spec.js:
```diff
@@ -13,7 +13,7 @@ describe("issue 30680", () => {
       .click();
 
-    cy.get(".ace_editor").type("select * from orders ");
+    cy.get(".ace_editor").type("select * from orders ", { delay: 100 });
```
unrelated to my changes but this was failing consistently on my branch for some reason 🤷‍♂️

sloansparger commented on e2e/test/scenarios/question/reproductions/17514-ui-overlay.cy.spec.js:
```diff
@@ -149,10 +150,15 @@ describe("issue 17514", () => {
 
       // Cypress cannot click elements that are blocked by an overlay so this will immediately fail if the issue is not fixed
-      cy.findByTextEnsureVisible("Subtotal").click();
+      openColumnOptions("Subtotal");
```
also unrelated to my changes and also consistently failing for some reason 🤷‍♂️

sloansparger commented on frontend/src/metabase/setup/components/LicenseTokenStep/LicenseTokenStep.tsx:
```diff
@@ -14,20 +14,22 @@ import type { NumberedStepProps } from "../types";
+    } catch (err) {
+      console.error(err);
+      throw new Error(
+        t`This token doesn’t seem to be valid. Double-check it, then contact support if you think it should be working.`,
```
There will be follow up work to make this submission provide better error messaging based on the work done by: https://github.com/metabase/metabase/pull/39067/

sloansparger committed:
resolve issues in license token tests

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

sloansparger committed:
removes token check endpoint

escherize committed:
remove unused var

sloansparger committed:
makes db dropdown disappear before typing in 30680 e2e test

npfitz reviewed (approved):
Gave it a test, and on the surface I could hardly tell things have changed haha. Good work!

johnswanson committed:
Default `anon-tracking-enabled` to false

I also went ahead and deleted the relevant test case, that previously
asserted that we would set `anon-tracking-enabled` to the value passed
to `/api/setup`. Since we don't support setting that via that API
endpoint anymore, having the test case lying around felt confusing.

johnswanson reviewed (approved):
Backend changes look good to me!

sloansparger committed:
Merge branch 'master' into ss/multi-req-setup

iethree commented on frontend/src/metabase/setup/actions.ts:
```diff
@@ -93,9 +92,28 @@ export const updateLocale = createAsyncThunk(
+      return rejectWithValue(error);
+    }
+
+    MetabaseSettings.set("setup-token", null);
```
(seeing this was just moved from another place) 👍

sloansparger committed:
revert c54c054e100610a2e60af62dfd790966f849d180

sloansparger commented on frontend/src/metabase/setup/tests/enterprise.unit.spec.tsx:
```diff
@@ -36,6 +36,10 @@ const setupEnterprise = (opts?: SetupOpts) => {
 const sampleToken = "a".repeat(64);
 
 describe("setup (EE, no token)", () => {
+  beforeEach(() => {
+    fetchMock.reset();
+  });
```
changing the order of the tests was causing a previously existing/passing test to fail. this was due to the new test mocking out fetch differently. ideally each test should be isolated, so i feel like we might should be doing this more places...

sloansparger committed:
pr feedback
